bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaAccountHandler.java

```diff
@@ -0,0 +1,550 @@
+        }
+    }
+
+    @Override
+    public void dispose() {
+        logger.trace("Disposing the Tesla account handler for {}", getThing().getUID());
+
+        lock.lock();
+        try {
+            if (connectJob != null && !connectJob.isCancelled()) {
+                connectJob.cancel(true);
+                connectJob = null;
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void scanForVehicles() {
+        scheduler.execute(() -> queryVehicles());
+    }
+
+    public void addVehicleListener(VehicleListener listener) {
+        this.vehicleListeners.add(listener);
+    }
+
+    public void removeVehicleListener(VehicleListener listener) {
+        this.vehicleListeners.remove(listener);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // we do not have any channels -> nothing to do here
+    }
+
+    public String getAuthHeader() {
+        if (logonToken != null) {
+            return "Bearer " + logonToken.access_token;
+        } else {
+            return null;
+        }
+    }
+
+    protected boolean checkResponse(Response response, boolean immediatelyFail) {
+        if (response != null && response.getStatus() == 200) {
+            return true;
+        } else {
+            apiIntervalErrors++;
+            if (immediatelyFail || apiIntervalErrors >= API_MAXIMUM_ERRORS_IN_INTERVAL) {
+                if (immediatelyFail) {
+                    logger.warn("Got an unsuccessful result, setting vehicle to offline and will try again");
+                } else {
+                    logger.warn("Reached the maximum number of errors ({}) for the current interval ({} seconds)",
+                            API_MAXIMUM_ERRORS_IN_INTERVAL, API_ERROR_INTERVAL_SECONDS);
+                }
+
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            } else if ((System.currentTimeMillis() - apiIntervalTimestamp) > 1000 * API_ERROR_INTERVAL_SECONDS) {
+                logger.trace("Resetting the error counter. ({} errors in the last interval)", apiIntervalErrors);
+                apiIntervalTimestamp = System.currentTimeMillis();
+                apiIntervalErrors = 0;
+            }
+        }
+
+        return false;
+    }
+
+    protected Vehicle[] queryVehicles() {
+
+        String authHeader = getAuthHeader();
+
+        if (authHeader != null) {
+            // get a list of vehicles
+            Response response = vehiclesTarget.request(MediaType.APPLICATION_JSON_TYPE)
+                    .header("Authorization", authHeader).get();
+
+            logger.debug("Querying the vehicle: Response: {}:{}", response.getStatus(), response.getStatusInfo());
+
+            if (!checkResponse(response, true)) {
+                logger.error("An error occurred while querying the vehicle");
+                return null;
+            }
+
+            JsonParser parser = new JsonParser();
```

Comment:
Why don't you use gson directly? Constructing a parser here looks very inefficient?

Reply:
This is old code I didn't touch, but creating a parser instance seems inefficient indeed.
But how would you do it with `gson` directly? Afaics, we do not have a pojo to parse the response to, but we only create a `JsonObject` for it - so we imho need a `JsonParser`. This could be moved to a field, though, to be reused.